File: 118.杨辉三角.py
#
# @lc app=leetcode.cn id=118 lang=python3
#
# [118] 杨辉三角
#

# @lc code=start
class Solution:
    def generate(self, numRows: int) -> List[List[int]]:
        res = []
        for i in range(1, numRows+1):
            res.append(self.getNum(i))
        return res
    
    def getNum(self, row: int) -> List[int]:
        row_num = [1]
        row = row - 1
        if row == 0:
            return row_num
        for i in range(1, int(row/2)+1):
            row_num.append(int(row_num[i-1]*(row-i+1)/i))
        if row%2 == 1:
            return row_num + row_num[::-1]
        elif row%2 == 0:
            return row_num + row_num[len(row_num)-2::-1]


# @lc code=end

File: 118.杨辉三角.java
import java.util.ArrayList;
import java.util.Collection;
import java.util.Collections;
import java.util.List;

/*
 * @lc app=leetcode.cn id=118 lang=java
 *
 * [118] 杨辉三角
 */

// @lc code=start
class Solution {
    public List<List<Integer>> generate(int numRows) {
        List<List<Integer>> res = new ArrayList<>();
        for(int i=0; i<numRows; i++){
            List<Integer> row = new ArrayList<>();
            row = getNum(i);
            res.add(row);
        }
        return res;
    }

    public List<Integer> getNum(int row){
        List<Integer> nums = new ArrayList<>();
        List<Integer> nums_reverse = new ArrayList<>();
        nums.add(1); 
        nums_reverse.add(0, 1);
        if(row==0){
            return nums;
        }
        for(int i=1; i<row/2+1; i++){
            int a = nums.get(i-1)*(row-i+1)/i;
            nums.add(a);
            nums_reverse.add(0, a);
        }
        if(row%2 == 0){
            nums_reverse.remove(0);
            nums.addAll(nums_reverse);
            return nums;
        }
        else{
            nums.addAll(nums_reverse);
            return nums;
        }
    }
}
// @lc code=end

